pkg/ipcutil/ipcutil.go

```diff
@@ -0,0 +1,250 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package ipcutil
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/oci"
+	"github.com/containerd/nerdctl/v2/pkg/idutil/containerwalker"
+	"github.com/containerd/nerdctl/v2/pkg/labels"
+	"github.com/docker/go-units"
+	"github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type IPCMode string
+
+type IPC struct {
+	Mode IPCMode `json:"mode,omitempty"`
+	// VictimContainer is only used when mode is container
+	VictimContainerID *string `json:"victimContainerId,omitempty"`
+
+	// HostShmPath is only used when mode is shareable
+	HostShmPath *string `json:"hostShmPath,omitempty"`
+
+	// ShmSize is only used when mode is private or shareable
+	// Devshm size in bytes
+	ShmSize string `json:"shmSize,omitempty"`
+}
+
+const (
+	Private   IPCMode = "private"
+	Host      IPCMode = "host"
+	Shareable IPCMode = "shareable"
+	Container IPCMode = "container"
+)
+
+// DetectFlags detects IPC mode from the given ipc string and shmSize string.
+// If ipc is empty, it returns IPC{Mode: Private}.
+func DetectFlags(ctx context.Context, client *containerd.Client, stateDir string, ipc string, shmSize string) (IPC, error) {
+	var res IPC
+	res.ShmSize = shmSize
+	switch ipc {
+	case "", "private":
+		res.Mode = Private
+	case "host":
+		res.Mode = Host
+	case "shareable":
+		res.Mode = Shareable
+		shmPath := filepath.Join(stateDir, "shm")
+		res.HostShmPath = &shmPath
+	default: // container:<id|name>
+		res.Mode = Container
+		parsed := strings.Split(ipc, ":")
+		if len(parsed) < 2 || parsed[0] != "container" {
+			return res, fmt.Errorf("invalid ipc namespace. Set --ipc=[host|container:<name|id>")
+		}
+
+		containerName := parsed[1]
+		walker := &containerwalker.ContainerWalker{
```

Comment:
this doesn't need to be a pointer?

Reply:
No, it does not have to. But does it have to be a value?
I think it is a very common usecase of `ContainerWalker`.

https://github.com/search?q=repo%3Acontainerd%2Fnerdctl%20ContainerWalker&type=code